content/blog/writing-crossguard-policies/index.md

```diff
@@ -0,0 +1,167 @@
+        // Then, check the buckets for violations.
+        for (const bucket of buckets) {
+            // Skip any buckets that haven't yet been provisioned (i.e. during previews).
+            if (!bucket.id) {
+                continue;
+            }
+            // If the bucket doesn't have any loggings and the bucket itself isn't being used as
+            // a log target, it's in violation of the policy.
+            if (!bucket.loggings || bucket.loggings.length === 0) {
+                if (!logBucketIDs.has(bucket.id)) {
+                    reportViolation("Bucket logging must be defined.", bucket.urn);
+                }
+            }
+        }
+    }),
+};
+
+// Utility method for defining a new StackValidation that will return the resources matching the provided type.
+function validateTypedResources<TResource extends Resource>(
+    resourceClass: { new(...rest: any[]): TResource },
+    validate: (
+        resources: q.Resolved<TResource>[],
+        args: StackValidationArgs,
+        reportViolation: ReportViolation) => Promise<void> | void,
+): StackValidation {
+    return (args: StackValidationArgs, reportViolation: ReportViolation) => {
+        const isInstance = (<any>resourceClass).isInstance;
+        if (!isInstance || typeof isInstance !== "function") {
+            return;
+        }
+        const resources = args.resources
+            .filter(r => isInstance({ __pulumiType: r.type }) === true)
+            .map(r => ({ ...r.props, urn: r.urn } as q.Resolved<TResource>));
+        validate(resources, args, reportViolation);
+    };
+}
+```
+
+The above example checks whether logging is enabled on S3 buckets. If a bucket does not have a `loggings` property defined, it will be in violation of the policy unless the bucket is being used as a logging `targetBucket` of another bucket. The policy check is skipped if the bucket's `id` isn't available, meaning this policy will only run during updates when the resource `id`s are available.
+
+At the time of writing, there is no built-in helper for stack validation policies similar to `validateTypedResource` for resource validation policies, so we've included our own helper that can be used until a [built-in helper](https://github.com/pulumi/pulumi-policy/issues/158) is available.
+
+## Resource Validation vs. Stack Validation
+
+|                                | Resource Validation                   | Stack Validation                                     |
+|--------------------------------|---------------------------------------|------------------------------------------------------|
+| What does it check?            | Individual resources                  | All resources in the stack                           |
+| When is the check performed?   | Before resources are created/modified | After all stack resources have been created/modified |
+| What information is available? | Resource _input_ properties           | Resource _output_ properties                         |
```

Comment:
For Stack Validation both input and output are available correct? Maybe this is implicit but i dont think its super clear

Reply:
It's just the outputs. https://github.com/pulumi/pulumi/blob/10a960ea4b7728a21fe0b0a6aa9768994a538e92/pkg/resource/deploy/step_generator.go#L1251-L1253

Although, during previews, inputs are propagated to outputs (as of https://github.com/pulumi/pulumi/pull/3245).

I'll add an explanation of the difference between previews and updates.